src/sidebar/services/frame-sync.js

```diff
@@ -134,86 +135,99 @@ export class FrameSyncService {
+        inSidebar.add(annot.$tag);
+        if (!this._inFrame.has(annot.$tag)) {
+          added.push(annot);
+        }
+      });
+      const deleted = prevAnnotations.filter(
+        annot => !inSidebar.has(annot.$tag)
+      );
+
+      // Send added annotations to matching frame.
+      if (added.length > 0) {
+        /** @type {Map<string|null, Annotation[]>} */
+        const addedByFrame = new Map();
+        for (let annotation of added) {
+          const frame = frameForAnnotation(frames, annotation);
+          if (!frame) {
+            continue;
           }
+          const anns = addedByFrame.get(frame.id) ?? [];
+          anns.push(annotation);
+          addedByFrame.set(frame.id, anns);
+        }
 
-          for (let [frameId, anns] of addedByFrame) {
-            const rpc = this._guestRPC.get(frameId);
-            if (rpc) {
-              rpc.call('loadAnnotations', anns.map(formatAnnot));
-            }
+        for (let [frameId, anns] of addedByFrame) {
+          const rpc = this._guestRPC.get(frameId);
+          if (rpc) {
+            rpc.call('loadAnnotations', anns.map(formatAnnot));
           }
-
-          added.forEach(annot => {
-            this._inFrame.add(annot.$tag);
-          });
         }
 
-        // Remove deleted annotations from frames.
-        deleted.forEach(annot => {
-          // Delete from all frames. If a guest is not displaying a particular
-          // annotation, it will just ignore the request.
-          this._guestRPC.forEach(rpc =>
-            rpc.call('deleteAnnotation', annot.$tag)
-          );
-          this._inFrame.delete(annot.$tag);
+        added.forEach(annot => {
+          this._inFrame.add(annot.$tag);
         });
+      }
+
+      // Remove deleted annotations from frames.
+      deleted.forEach(annot => {
+        // Delete from all frames. If a guest is not displaying a particular
+        // annotation, it will just ignore the request.
+        this._guestRPC.forEach(rpc => rpc.call('deleteAnnotation', annot.$tag));
+        this._inFrame.delete(annot.$tag);
+      });
 
-        // Update elements in host page which display annotation counts.
-        if (frames.length > 0) {
-          if (frames.every(frame => frame.isAnnotationFetchComplete)) {
-            if (publicAnns === 0 || publicAnns !== prevPublicAnns) {
-              this._hostRPC.call('publicAnnotationCountChanged', publicAnns);
-              prevPublicAnns = publicAnns;
-            }
+      // Update elements in host page which display annotation counts.
+      if (frames.length > 0) {
+        if (frames.every(frame => frame.isAnnotationFetchComplete)) {
+          if (publicAnns === 0 || publicAnns !== prevPublicAnns) {
+            this._hostRPC.call('publicAnnotationCountChanged', publicAnns);
+            prevPublicAnns = publicAnns;
           }
         }
       }
+    };
+
+    watch(
+      this._store.subscribe,
+      () =>
+        /** @type {const} */ ([
```

Comment:
Eh?

Reply:
This is a [const assertion](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions), which has the effect of making TypeScript infer the type as a fixed-sized tuple (here `[Annotation[], Frame[]]`) rather than a more generic array (`Array<Annotation|Frame>`). This results in the `annotations`, `frames` and `prevAnnotations` arguments below all having the expected types.